server/user.go

```diff
@@ -295,18 +313,30 @@ func (s *UserMgmtServer) getCreds(r *http.Request, requiresAdmin bool) (api.Cred
 		return api.Creds{}, err
 	}
 
-	isAdmin, err := s.cm.IsDexAdmin(clientID)
-	if err != nil {
-		log.Errorf("userMgmtServer: GetCreds err: %q", err)
-		return api.Creds{}, err
+	i := 0
+	for _, clientID := range clientIDs {
+		// Make sure the client actually exists.
+		isAdmin, err := s.cm.IsDexAdmin(clientID)
+		if err != nil {
+			log.Errorf("userMgmtServer: GetCreds err: %q", err)
+			return api.Creds{}, err
+		}
+		// If the endpoint requires an admin client, filter out clients which are not admins.
+		if isAdmin || !requiresAdmin {
+			clientIDs[i] = clientID
+			i++
+		}
 	}
-	if requiresAdmin && !isAdmin {
+
+	clientIDs = clientIDs[:i]
```

Comment:
Hmm that's what I suspected. Sorry if I misunderstood, I just want to
clarify what I meant:  what if all the clients except for one are
admins and all but one (the non-admin) are expired -> then 278 returns ok.
Afterwards, on this line the clients would be trimmed down to hold only the
admins (which happens to be all expired clients). Then the splice of
clients that is left holds only invalid admins. Would these clients be
given access to endpoints they should not have access to? Let me know
if this is not actually an issue.

On Monday, August 1, 2016, Eric Chiang <notifications@github.com> wrote:

> In server/user.go
> <https://github.com/coreos/dex/pull/531#discussion_r73090912>:
>
> >  	}
> > -	if requiresAdmin && !isAdmin {
> > +
> > +	clientIDs = clientIDs[:i]
>
> The check on line 278 is verifying the token is signed by dex, has the
> correct issuer, isn't expired, etc. This is a bit different, it's checking
> if the client itself has "admin" status. I think agree that the logic could
> be cleaned up a bit.
>
> So maybe, a better comment on line 278, and more obvious logic in this
> chunk?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coreos/dex/pull/531/files/57c58184dff08fd071686d9268ee8aff13d1d280#r73090912>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATiQP3zpMeE76PpdJL-xAffFq-GVrhoZks5qbr8QgaJpZM4JZ_4s>
> .
>